### DISCUSSION TITLE
Circuit-relay v2 connection denied (`Upgrade(Apply(NoReservation))`)

### DISCUSSION THREAD

**Discussion by demfabris** [libp2p/rust-libp2p#2513] [Q&A] [answered]
I'm facing problems when establishing a relayed connection using the recently implemented circuit-relay-v2

## Scenario
**Node A**, **R**, and **B**

1. **Node R** starts a p2p node supporting `Identify, circuit-relay-v2/hop, circuit-relay-v2/stop`
code:
```shell
$ mynode --listen 127.0.0.1 --port 4001
```

2. **Node B** starts a p2p node supporting the same behaviours and dials **Node R**, establishing a connection successfully.
```shell
$ mynode --listen 127.0.0.1 --port 4002
$ mynode --dial /ip4/127.0.0.1/tcp/4001/p2p/<NodeRpeerId> 
```

3. **Node A** starts a p2p node supporting the same behaviours and attempts to dial **Node B** via **Node R** as relay.
```shell
$ mynode --listen 127.0.0.1 --port 4003
$ mynode --dial /ip4/127.0.0.1/tcp/4001/p2p/<NodeRpeerId>/p2p-circuit/p2p/<NodeBpeerId>
```

## Result
**Node A** establishes a connection with **Node R** successfully, but then a error follows in both of them:

**Node R** output:
```text
ERROR_LOG: CircuitReqDenied { src_peer_id: PeerId(<nodeA>), dst_peer_id: PeerId(<nodeB>) }
```

**Node A** output:
```text
ERROR_LOG: Behaviour(Client(OutboundCircuitReqFailed { relay_peer_id: PeerId(<nodeR>), error: Upgrade(Apply(NoReservation)) }))

ERROR_LOG: Transport([("/ip4/127.0.0.1/tcp/4001/p2p/<nodeR>/p2p-circuit/p2p/<nodeB>", Other(Custom { kind: Other, error: Other(A(A(A(A(B(Connect)))))) }))])
```

**Node B** output:
Nothing

What step I might be missing here?

**Comment by mxinden** [comment, marked as answer]
:wave: sorry for the late reply here.

It looks like Node B does not make a reservation with R. You would need to make sure Node B listens via R, in other words that node B makes a reservation with R. Unless B has a reservation with R, R will not relay a connection from A to B.

To "listen via" R, B needs to call `Swarm::listen_on` with a relayed address, e.g. `/ip4/127.0.0.1/tcp/4001/p2p/<NodeRpeerId>/p2p-circuit`.

See example:

https://github.com/libp2p/rust-libp2p/blob/e66f04f41b35aed7d3bf287e5d0cbd82987587fa/protocols/relay/tests/v2.rs#L53-L60

Does that solve your issue @demfabris?

**Reply by demfabris** [reply]
That's exactly what I was missing 😄 
This wasted me a couple of days... Not very much disclosed in the docs.

Thanks

**Reply by mxinden** [reply]
>  This wasted me a couple of days... Not very much disclosed in the docs.

Sorry to hear. Would you mind proposing a docs change (PR) to avoid future users running in the same issue?
